### PR TITLE
[Search-based access requests] Implement "tsh ssh --request user@node"

### PR DESCRIPTION
**Pull request by nklaassen**
See [RFD 59](https://github.com/gravitational/teleport/blob/master/rfd/0059-search-based-access-requests.md) for background.

This PR implements the new flag `tsh ssh --request`. When this flag is set, `tsh` will attempt to automatically request access to the specific SSH node the user is trying to access using a search-based access request.

## Demo
```bash
$ tsh ssh nklaassen@one-node
ERROR: access denied to nklaassen connecting to one-node on cluster cluster-one

$ tsh ssh --request nklaassen@one-node
ERROR: access denied to nklaassen connecting to one-node on cluster cluster-one

You do not currently have access to nklaassen@one-node, attempting to request access.

Enter request reason (Ctrl-C to cancel): responding to incident 123
Creating request...
Request ID: 6f93eb80-32a0-4e51-887e-b7bb0691875f
Username:   nic
Roles:      access
Resources:  [{"cluster":"cluster-one","kind":"node","name":"bbb56211-7b54-4f9e-bee9-b68ea156be5f"}]
Reason:     "responding to incident 123"
Reviewers:  [none] (suggested)
Status:     PENDING

hint: use 'tsh login --request-id=<request-id>' to login with an approved request

Waiting for request approval...

Approval received, getting updated certificates...

one-node$ date
Tue 31 May 2022 17:42:34 PDT
one-node$ exit
logout
the connection was closed on the remote side on  31 May 22 17:42 PDT
$
```

## TODO
Optimize the set of roles which are automatically requested. Currently it will be all `search_as_roles` configured for the user, it should be possible to get it down to 1 single role which has access to the node with the requested username.

### CODE REVIEW

File: tool/tsh/tsh.go [hunk 1]
```diff
@@ -2063,6 +2073,138 @@ func serializeClusters(rootCluster clusterInfo, leafClusters []clusterInfo, form
...
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	}
+	req.SetRequestReason(requestReason)
+
+	// Watch for resolution events on the given request. Start watcher and wait
+	// for it to be ready before creating the request to avoid a potential race.
+	requestWatcher := newAccessRequestWatcher(req)
+	defer requestWatcher.Close()
```
[the remark this answers is not in the document]
Comment (nklaassen) [reply tool/tsh/tsh.go:2164]: We handle it in the happy case a few lines below (`Close()` should always be safe to call multiple times). The deferred `Close()` is really convenient to handle all the error cases in between where `tsh` is just going to die anyway, I don't think checking the `Close()` error is that important there, most of the time we don't do it at all.

File: tool/tsh/tsh.go [hunk 2]
```diff
@@ -2871,60 +3023,113 @@ func host(in string) string {
...
-// approval, returns the updated request. `clt` must be a client to the root
-// cluster, such as the one returned by
-// `(*TeleportClient).WithRootClusterClient`. `ready` will be closed when the
-// event watcher used to wait for the request updates is ready.
-func waitForRequestResolution(cf *CLIConf, clt auth.ClientI, req types.AccessRequest, ready chan<- struct{}) (types.AccessRequest, error) {
+// accessRequestWatcher is a helper to wait for an access request to be resolved.
+type accessRequestWatcher struct {
+	req     types.AccessRequest
+	watcher types.Watcher
+	closers []io.Closer
```
[the remark this answers is not in the document]
Comment (nklaassen) [reply tool/tsh/tsh.go:3030]: Why? There's no concurrent access. I suppose it's not safe to call `awaitResolution` or `Close` concurrently with `initialize`, but I don't claim that `accessRequestWatcher` is safe for concurrent use and its design is meant to eliminate the need for concurrency.

File: tool/tsh/tsh_test.go [hunk 3]
```diff
@@ -490,6 +491,192 @@ func TestMakeClient(t *testing.T) {
...
+			select {
+			case <-stop:
+				return
+			case <-ticker:
+			}
+			requests, err := access.GetAccessRequests(ctx,
+				types.AccessRequestFilter{
+					State: types.RequestState_PENDING,
+				})
+			require.NoError(t, err)
```
[the remark this answers is not in the document]
Comment (nklaassen) [reply tool/tsh/tsh_test.go]: Oh thanks, wasn't thinking about that, should be fixed now

File: tool/tsh/tsh.go [hunk 4]
```diff
@@ -2071,38 +2213,48 @@ func onSSH(cf *CLIConf) error {
...
-			for _, node := range allNodes {
-				if node.GetHostname() == tc.Host {
-					nodes = append(nodes, node)
+	err = retryWithAccessRequest(cf, tc, func() error {
+		err = client.RetryWithRelogin(cf.Context, tc, func() error {
+			return tc.SSH(cf.Context, cf.RemoteCommand, cf.LocalExec)
+		})
+		if err != nil {
+			if strings.Contains(utils.UserMessageFromError(err), teleport.NodeIsAmbiguous) {
+				allNodes, err := tc.ListAllNodes(cf.Context)
```
[the remark this answers is not in the document]
Comment (nklaassen) [reply tool/tsh/tsh.go]: yeah, this is just what was there, but I went ahead and updated it to use the predicate search.

File: tool/tsh/tsh.go [hunk 5]
```diff
@@ -2063,6 +2073,138 @@ func serializeClusters(rootCluster clusterInfo, leafClusters []clusterInfo, form
...
+		return nil, trace.Wrap(err)
+	}
+	defer proxyClient.Close()
+
+	// First search by hostname, then host ID
+	filter := proto.ListResourcesRequest{
+		UseSearchAsRoles: true,
+	}
+	hostNameExpr := fmt.Sprintf(`resource.spec.hostname == "%s"`, tc.Host)
+	hostIDExpr := fmt.Sprintf(`name == "%s"`, tc.Host)
```
[the remark this answers is not in the document]
Comment (nklaassen) [reply tool/tsh/tsh.go]: I'm not sure, I don't see `resource.spec.hostname` used anywhere else, I think it's too early to start making a collection of common templates when they aren't used much yet

File: tool/tsh/tsh_test.go [hunk 6]
```diff
@@ -490,6 +492,196 @@ func TestMakeClient(t *testing.T) {
...
+		"echo", "test",
+	}, setHomePath(tmpHomePath))
+	require.NoError(t, err)
+
+	// now that we have an approved access request, it should still work without --request
+	err = Run([]string{
+		"ssh",
+		"--insecure",
+		"--debug",
+		"--request-reason", "for testing purposes",
```
[the remark this answers is not in the document]
Comment (nklaassen) [reply tool/tsh/tsh_test.go:648]: this just avoids the prompt tsh will try to use to fill in the reason if it's omitted

**Comment by nklaassen** [comment]
Thanks for the reviews @rosstimothy, should be ready for another pass. Friendly ping @EdwardDowling 🙏

**Review by rosstimothy — APPROVED** [review]
👍

**Review by EdwardDowling — APPROVED** [review]

**Pull request merged**